### PR TITLE
Use Y-m-d format for creation dates in extension XML. Adds date sorting to extension manager and localised date formats.

### PR DESCRIPTION
**Pull request by Bakual**
The translators decided to standardise the date format in their language packs to use the ISO 8601 date format. Thus it makes sense that core uses the same format for core language XML as well.

The build script currently only knows one creation date for all XML files, so it will affect also `administrator/manifests/files/joomla.xml`, not only language ones.

Initially this PR only changed the build script and the XMLs updated by it. I've expanded it now so it contains all core XML files.

Also in the extension manager, you can sort by date and the date is formatted according to the active language.

### Summary of Changes
Changing the format of the creation date to use eg 2019-11-26 instead of November 2019

### Testing Instructions
You can try and run the build/bump.php and see the creation date in the XML files. That script is used when a new release is built.

administrator/language/en-GB/en-GB.xml
administrator/language/en-GB/install.xml
administrator/manifests/files/joomla.xml
administrator/manifests/packages/pkg_en-GB.xml
installation/language/en-GB/en-GB.xml
language/en-GB/en-GB.xml
language/en-GB/install.xml

In the extension manager, try sorting by date.
It may be necessary to select all extensions and hit the "Refresh Cache" button so the creation dates are updated from the XMLs.
Also try with different languages and see if the date format changes.

### Expected result
Date format like 2019-11-26 (or 26.11.2019 or whatever is specific to the active language)

### Actual result
Date format like November 2019

### Documentation Changes Required
None

### CODE REVIEW

**Comment by brianteeman** [comment]
Do we really need to know the day?

**Comment by brianteeman** [comment]
Also it introduces an inconsistency as all the other files use Month Year as their creationDate.

image

**Comment by Bakual** [comment]
Personally I don't care much how it is done. But I see indeed advantages in it.
* I think the day is part of a date and should be there. I don't see a reason to not show it, especially since it is part of a script.
* The current way is using english words which aren't translated. The proposed way would be language independent. It could even be translated into different formats and used to sort the list, if we ever want to do that (and get extensions to consistently use it).

**Comment by brianteeman** [comment]
I understand and appreciate the issue with it being in English. However for me the problem is 
1. Consistency as explained above
2. Clarity - is 2019-1-11 the first of November or the eleventh of January.

**Comment by ot2sen** [comment]
2 is exactly the reason for the ISO standard to get rid of that confusion
og which is which. Y-M-D :)
Good call by the translators.
Would be fine if the older are updated for consistency

**Comment by brianteeman** [comment]
Try explains that to the people who live in the USA

**Comment by alikon** [comment]
why to have 2 format ? 
one for `	'reldate'    => date('j-F-Y'),`  
and  another one for `'credate'    => date('Y-m-d'),`

**Comment by infograf768** [comment]
I am a bit confused.
How would such dates be translated in the language packs xmls as we can enter whatever we want there?

**Comment by Bakual** [comment]
> How would such dates be translated in the language packs xmls as we can enter whatever we want there?

The point is that you are not supposed to enter whatever you want. According to the forum post by Mig. :)
I'm not saying we're going to translate them. I just say it could be done if at least the majority would follow that format.

**Comment by brianteeman** [comment]
>  I just say it could be done if at least the majority would follow that format.

and all of the core xml would need to be updated (they are not covered by the bump script) and I still say it is more confusing to have 2019-1-11

**Comment by Bakual** [comment]
> I still say it is more confusing to have 2019-1-11

The format Y-d-m isn't used in that many countries. According to Wikipedia it's only 4 countries. So my guess would be most would read that date as January 11, 2019.

Rewriting all the other core XMLs would be rather simple, I can do that of course. My initial goal was just the language pack files so they follow what the translators have as their guidelines.

**Comment by ot2sen** [comment]
Would be nice to have it fully covered. The English text currently from the
XMLs are only of use to some while the ISO standard is of use to everyone :)

**Comment by brianteeman** [comment]
Might only be 4 countries but it's a significant % of users

**Comment by Bakual** [comment]
> Might only be 4 countries but it's a significant % of users

I honestly don't know how that equates to user. Countries are Kazakhstan, Latvia, Nepal and Turkmenistan.
On the other hand, using an english word for the month isn't that inclusive either.

I will expand this PR so it includes all created dates. And I'll have a look at how work it would be to make it translateable and sortable.

**Comment by laoneo** [comment]
Using an ISO standard is a good move. Since today I'm using the day also in my releases. If accepted extension devs can follow it.

**Comment by Bakual** [comment]
I've expanded this PR so it changes all `creationDate` instances to the ISO format.
I've also added sorting by date to the extension manager and the date will be shown "translated" where possible (depending on the source format of the date).

Obviously, sorting by date doesn't work properly for the dates that aren't entered by the ISO format. This gives a gentle push to extension devs to also use the ISO format for the date.

**Comment by brianteeman** [comment]
I stand by the comment that this will cause confusion as people wont know what month 2019-11-1 is

**Comment by Bakual** [comment]
> I stand by the comment that this will cause confusion as people wont know what month 2019-11-1 is

The date is now shown localised in the format by the active language. So no confusion anymore.

**Comment by Bakual** [comment]
english:
english
german:
deutsch

**Comment by brianteeman** [comment]
You are assuming that everyone has the correct language pack installed. For example there have only been 6000 downloads of the US-English

**Comment by rdeutz** [comment]
Test needs to be fixed when we go this way:

```
1) JInstallerManifestLibraryTest::testLoadManifestFromData

Failed asserting that two strings are equal.

--- Expected

+++ Actual

@@ @@

-'2008'

+'2008-01'
```

**Comment by Bakual** [comment]
> You are assuming that everyone has the correct language pack installed.

No, I'm not assuming this.
But if people would really get confused with the date format, then there would be an easy solution available by installing and selecting the language.

> Test needs to be fixed when we go this way:

Test fixed.

**Comment by Bakual** [comment]
Rebased and conflicts solved

**Review by laoneo — APPROVED** [review]

**Comment by N6REJ** [comment]
I have tested this item :red_circle: unsuccessfully on f9ed44be0c09db2092b25a1c938673a497969991<br /><br />Didn't make any difference after applying the patch via patchtester.  Nor did it respect the currently language after I set us english as default.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27162">issues.joomla.org/tracker/joomla-cms/27162</a>.</sub>

image
image

**Comment by Bakual** [comment]
That the date format didn't change with US English is actually due to the US Englisch language pack being faulty. The respective language string is `DATE_FORMAT_LC4="Y-m-d"`. Same as in en-GB. I think in US you use "Y-d-m" instead?
If you adjust that string in the en-US.ini file (or by using a language override) it should work.

Also, it may be necessary to select all extensions and hit the "Refresh Cache" button so the creation dates are updated from the XMLs.

**Comment by chmst** [comment]
Using the patchtester I got the message: 
The file marked for modification does not exist: administrator/components/com_installer/models/extension.php

**Comment by Quy** [comment]
@chmst This PR is for J3 and not J4.

**Comment by N6REJ** [comment]
> That the date format didn't change with US English is actually due to the US Englisch language pack being faulty. The respective language string is `DATE_FORMAT_LC4="Y-m-d"`. Same as in en-GB. I think in US you use "Y-d-m" instead?

nope, we use m-d-Y

**Comment by N6REJ** [comment]
> Also, it may be necessary to select all extensions and hit the "Refresh Cache" button so the creation dates are updated from the XMLs.

shouldn't the "january 2020" have changed to numeric?

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on f9ed44be0c09db2092b25a1c938673a497969991<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27162">issues.joomla.org/tracker/joomla-cms/27162</a>.</sub>

**Comment by infograf768** [comment]
@N6REJ 
There is no more "january 2020" after patch.

**Comment by alikon** [comment]
i've not runned the bump script
when i sort let's say for date descending   the result is incorrect
Screenshot from 2020-02-15 08-30-50

**Comment by infograf768** [comment]
> when i sort let's say for date descending the result is incorrect

I must say I did not test that.

**Comment by brianteeman** [comment]
probably because the date is being sorted by the stored value and not the displayed value

**Comment by Bakual** [comment]
Make sure to select all extensions and use the "Refresh Cache" button. That is because the dates from the XMLs are stored in the database. When I changed the XMLs, the database isn't automatically updated and thus still has the old values.
I think the database is updated during the update process, but honestly I'm not sure about that. Maybe someone else knows.

**Comment by infograf768** [comment]
Sorting by date works here (I had refreshed cache).
Evidently, if the xml has not been modified (for example patchtester or Installation from web), we have weird results.

**Comment by Bakual** [comment]
Merged current staging and resolved conflicts

**Comment by Bakual** [comment]
Is there any interest in getting this into core? If so, I can update the branch, otherwise I would close it.

**Comment by laoneo** [comment]
Perhaps more for 4.1?

**Comment by Bakual** [comment]
Sure, for 4.0 it's to late 😄

**Comment by laoneo** [comment]
Would be nice to rebase for 4.1 and then get another round of testing.

**Comment by HLeithner** [comment]
> Is there any interest in getting this into core? If so, I can update the branch, otherwise I would close it.

Yes, please update the PR for 4.1.

**Comment by Bakual** [comment]
I've tried on my laptop, but messed it up somehow with the conflict resolving. I need to do it when I'm back home.

**Comment by Bakual** [comment]
Recreated for 4.1 as there were two many changes between 3.10 and 4.1. It was easier to cherrypick the things I need and redo the rest manually.
I couldn't test it yet myself but it should work the same as before.

**Comment by laoneo** [comment]
@Bakual can you fix the conflicts and then hopefully we will find testers for this one as it is a nice change.

**Comment by Bakual** [comment]
Rebased to 4.2 and solved conflicts.

**Comment by Bakual** [comment]
Found a few more creationDates. Should be ready for testin again.

**Comment by laoneo** [comment]
Thanks. There are still some conflicts in installation and api.

**Comment by Bakual** [comment]
Richard promised me they will be gone once 4.1 is merged up into 4.2 next time.

File: plugins/editors/tinymce/tinymce.xml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <extension type="plugin" group="editors" method="upgrade">
 	<name>plg_editors_tinymce</name>
 	<version>5.10.2</version>
-	<creationDate>2005-2021</creationDate>
+	<creationDate>2005-01</creationDate>
```
Comment (brianteeman) [inline plugins/editors/tinymce/tinymce.xml]: Clearly this is wrong

Comment (Bakual) [reply plugins/editors/tinymce/tinymce.xml]: Nope

Comment (brianteeman) [reply plugins/editors/tinymce/tinymce.xml]: How is "2005 to 2021" the same as "January 2005"

or How can it be january 2005 if joomla didnt start until august 2005

Comment (Bakual) [reply plugins/editors/tinymce/tinymce.xml]: I didn't say it's the same, but it's not wrong 😄 
If you feel better, I can change it to 2005-08. But it has to be a date and not a range.

**Comment by Bakual** [comment]
Tried again and changed the 2005-01 date to 2005-08 so it matches the Joomla release

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on ea91a81645e8e63be5ad0d461936f3b194d697bb<br /><br />I could not install a language as they do not exist for 4.2 but I did a language override of the constant DATE_FORMAT_LC4 and then the date in the extension manager has changed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27162">issues.joomla.org/tracker/joomla-cms/27162</a>.</sub>

**Comment by richard67** [comment]
@Bakual Is the description of this PR still right? It says:

> Initially this PR only changed the build script and the XMLs updated by it. I've expanded it now so it contains all core XML files.

But in the changes I cannot find any change in the build script.

**Comment by Bakual** [comment]
Looks like it got lost during the several redoing/conflict solving of this PR.

**Comment by Bakual** [comment]
Added back the changes in the bump script.

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on ae3da4835dfb3f1c2ae28c3e4ba0f9bcd1b4021f<br /><br />I tested it again with a language override of the constant DATE_FORMAT_LC4 and then the date in the extension manager has changed. Run also the buzmp script `php build/bump.php -v 4.2.9` and then the first entry showed version 4.2.9 with the current date.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27162">issues.joomla.org/tracker/joomla-cms/27162</a>.</sub>

**Comment by laoneo** [comment]
The first entries are listed from 2004, is this correct?

**Comment by Bakual** [comment]
> The first entries are listed from 2004, is this correct?

I didn't validate the dates. I just took what was there and changed the format. 😄

**Comment by richard67** [comment]
> Run also the buzmp script `php build/bump.php -v 4.2.9` and then the first entry showed version 4.2.99 with the current date.

@laoneo I hope this is a typo and the patch version was not changed from 9 to 99.

**Comment by laoneo** [comment]
Yes sure was a typo, just wanted to make sure it used my bump version.

**Comment by Bakual** [comment]
Merged upstream and solved conflicts

**Comment by roland-d** [comment]
Thanks everybody.

**Comment by Bakual** [comment]
Wohoooo! 🕺🏻

**Comment by laoneo** [comment]
Nice!!

**Comment by korenevskiy** [comment]
It's obvious. An XML file is a data storage format. I repeat this is the data format. XML tags can store any type of data. If only the "CreationDate" tag is a text type, then the format of this tag is obvious, but if the "creation Date" tag is a date type, then this implies the format in ISO. If we remember that XML is a data format, and XML is not a text document of articles. XML is a data warehouse, just like SQL is a data warehouse. The difference is that SQL is data processing and storage, and XML is just data storage. 
But let's look at this problem from the other side. Third-party developers can create components for sorting by date tag. 
Even if there is no sorting in Joomla, the data storage format must match the content.
**We store data in SQL in ISO format. Maybe we will store data in SQL in the "Mov 2022" format?**

**Pull request merged**